File: src/commands/format.ts
import path from 'path';
import { highlightMentions } from '../core/mentions.js';
import { parseAgentId } from '../core/agents.js';

const NO_COLOR = !!process.env.NO_COLOR;

const DIM = NO_COLOR ? '' : '\x1b[2m';
const BOLD = NO_COLOR ? '' : '\x1b[1m';
const BLUE = NO_COLOR ? '' : '\x1b[34m';
const GRAY = NO_COLOR ? '' : '\x1b[38;5;240m';
const RESET = NO_COLOR ? '' : '\x1b[0m';
const ISSUE_BG = NO_COLOR ? '' : '\x1b[48;5;17m';
const WHITE_TEXT = NO_COLOR ? '' : '\x1b[38;5;231m';

const MAX_DISPLAY_LINES = 20;

const USER_COLOR = NO_COLOR ? '' : '\x1b[38;5;249m';

export const COLOR_PAIRS = [
  { bright: '\x1b[38;5;111m', dim: '\x1b[38;5;105m' },  // Blue
  { bright: '\x1b[38;5;157m', dim: '\x1b[38;5;156m' },  // Green
  { bright: '\x1b[38;5;216m', dim: '\x1b[38;5;215m' }, // Orange
  { bright: '\x1b[38;5;36m', dim: '\x1b[38;5;30m' },   // Teal
  { bright: '\x1b[38;5;183m', dim: '\x1b[38;5;141m' },  // Purple
  { bright: '\x1b[38;5;230m', dim: '\x1b[38;5;229m' }, // Yellow
];

/**
 * Simple hash function for strings.
 * Returns a non-negative integer.
 */
function hashString(str: string): number {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash = hash & hash; // Convert to 32-bit integer
  }
  return Math.abs(hash);
}

/**
 * Get ANSI color code for an agent name with optional color map.
 * - Users (type='user'): no color
 * - System messages (from_agent='system'): gray
 * - Agents: color based on map (if provided), or hash of base name
 * - Bright/dim based on version parity
 * @param agentId Agent ID (e.g., "alice.1")
 * @param type Message type ('agent' or 'user')
 * @param colorMap Optional map from agent base to color index (0-5)
 * @returns ANSI color code string
 */
export function getAgentColor(agentId: string, type: string = 'agent', colorMap?: Map<string, number>): string {
  if (NO_COLOR) {
    return '';
  }

  if (agentId === 'system') {
    return GRAY;
  }

  try {
    const parsed = parseAgentId(agentId);

    // Use color map if provided and base exists in map
    let colorIndex: number;
    if (colorMap && colorMap.has(parsed.base)) {
      colorIndex = colorMap.get(parsed.base)!;
    } else {
      // Fall back to hash-based assignment
      colorIndex = hashString(parsed.base) % COLOR_PAIRS.length;
    }

    const pair = COLOR_PAIRS[colorIndex];

    // Odd version = bright, even version = dim
    return parsed.version % 2 === 1 ? pair.bright : pair.dim;
  } catch {
    // Plain name without version (e.g., "adam") - treat as version 1
    const colorIndex = hashString(agentId) % COLOR_PAIRS.length;
    const pair = COLOR_PAIRS[colorIndex];
    return pair.bright;
  }
}

/**
 * Extract project name from project path.
 * Uses the directory name containing .mm/
 */
export function getProjectName(project: { root: string; dbPath: string }): string {
  return path.basename(project.root);
}

/**
 * Highlight issue IDs with dark blue background and white text.
 * Matches pattern: #<prefix>-<alphanumeric> (e.g., #mm-123, #myproject-5ar)
 * @param body Message body text
 * @param senderColor The sender's color code (used to restore after each highlight)
 * @returns Body with highlighted issue IDs
 */
function highlightIssueIds(body: string, senderColor: string = ''): string {
  return body.replace(/#([a-z][a-z0-9]*)-([a-z0-9]+)\b/gi, (match) => {
    return `${ISSUE_BG}${WHITE_TEXT}${match}${RESET}${senderColor}`;
  });
}

/**
 * Colorize @mentions in a message body.
 * Replaces @mentions with colored versions matching the mentioned agent.
 * Validates against known agent bases when provided.
 * @param body Message body text
 * @param senderColor The sender's color code (used to reset after each mention)
 * @param agentBases Optional set of known agent base names for validation
 * @returns Body with colorized mentions
 */
function colorizeBody(body: string, senderColor: string, agentBases?: Set<string>): string {
  return body.replace(/@([a-z][a-z0-9.]*)/gi, (match, name) => {
    const lowercase = name.toLowerCase();

    // Always colorize @all
    if (lowercase === 'all') {
      const mentionColor = getAgentColor(name, 'agent');
      if (mentionColor) {
        return `${mentionColor}${match}${senderColor}`;
      }
    }

    // Colorize if it has a dot (agent ID like alice.1)
    if (name.includes('.')) {
      const mentionColor = getAgentColor(name, 'agent');
      if (mentionColor) {
        return `${mentionColor}${match}${senderColor}`;
      }
    }

    // For plain names, validate against agent bases if provided
    if (agentBases) {
      if (agentBases.has(lowercase)) {
        const mentionColor = getAgentColor(name, 'agent');
        if (mentionColor) {
          return `${mentionColor}${match}${senderColor}`;
        }
      }

      // User mentions: not in agentBases, no dots, 3-15 chars
      if (!name.includes('.') && name.length >= 3 && name.length <= 15) {
        return `${BOLD}${match}${RESET}${senderColor}`;
      }
    } else {
      // No validation available - use fallback heuristic
      // Colorize names that are 3-15 chars (typical usernames)
      if (name.length >= 3 && name.length <= 15) {
        const mentionColor = getAgentColor(name, 'agent');
        if (mentionColor) {
          return `${mentionColor}${match}${senderColor}`;
        }
      }
    }

    return match;
  });
}

/**
 * Truncate message body for display if it exceeds MAX_DISPLAY_LINES.
 * Full message remains in DB, but display shows first N lines with indicator.
 */
function truncateForDisplay(body: string, msgId: string): string {
  const lines = body.split('\n');
  if (lines.length <= MAX_DISPLAY_LINES) {
    return body;
  }

  const truncated = lines.slice(0, MAX_DISPLAY_LINES).join('\n');
  const remaining = lines.length - MAX_DISPLAY_LINES;
  return truncated + '\n' +
         `... (${remaining} more lines. Use 'mm view ${msgId}' to see full)`;
}

/**
 * Format a message for display.
 * Format: [#projectname guid] @speaker: "message"
 * - Project/message GUID block is dimmed
 * - Project name (#projectname) is bold within the dim block
 * - Entire message body is colorized based on sender agent
 * - @mentions within body are colorized to match mentioned agent
 */
export function formatMessage(msg: any, projectName: string, agentBases?: Set<string>): string {
  const idBlock = `${DIM}[${BOLD}#${projectName}${RESET}${DIM} ${msg.id}]${RESET}`;

  const color = getAgentColor(msg.from_agent, msg.type);
  const truncatedBody = truncateForDisplay(msg.body, msg.id);

  if (color) {
    // Color entire message body including speaker and text
    // Apply @mention colorization first, then issue ID highlighting
    const colorizedBody = highlightIssueIds(colorizeBody(truncatedBody, color, agentBases), color);
    return `${idBlock} ${color}@${msg.from_agent}: "${colorizedBody}"${RESET}`;
  } else {
    // User messages - no color
    const body = highlightIssueIds(highlightMentions(truncatedBody));
    return `${idBlock} @${msg.from_agent}: "${body}"`;
  }
}

File: internal/command/format.go
package command

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"unicode"
	"unicode/utf8"

	"github.com/adamavenir/fray/internal/core"
	"github.com/adamavenir/fray/internal/types"
)

const maxDisplayLines = 20

// Accordion settings
const (
	DefaultAccordionThreshold = 10 // Show accordion if more than this many messages
	AccordionHeadCount        = 3  // Number of messages to show at start
	AccordionTailCount        = 3  // Number of messages to show at end
)

var (
	noColor = os.Getenv("NO_COLOR") != ""

	dim       = ansiCode("\x1b[2m")
	bold      = ansiCode("\x1b[1m")
	gray      = ansiCode("\x1b[38;5;240m")
	reset     = ansiCode("\x1b[0m")
	issueBG   = ansiCode("\x1b[48;5;17m")
	whiteText = ansiCode("\x1b[38;5;231m")
	cyan      = ansiCode("\x1b[36m")
)

var colorPairs = []struct {
	Bright string
	Dim    string
}{
	{Bright: ansiCode("\x1b[38;5;111m"), Dim: ansiCode("\x1b[38;5;105m")},
	{Bright: ansiCode("\x1b[38;5;157m"), Dim: ansiCode("\x1b[38;5;156m")},
	{Bright: ansiCode("\x1b[38;5;216m"), Dim: ansiCode("\x1b[38;5;215m")},
	{Bright: ansiCode("\x1b[38;5;36m"), Dim: ansiCode("\x1b[38;5;30m")},
	{Bright: ansiCode("\x1b[38;5;183m"), Dim: ansiCode("\x1b[38;5;141m")},
	{Bright: ansiCode("\x1b[38;5;230m"), Dim: ansiCode("\x1b[38;5;229m")},
}

var (
	mentionRe = regexp.MustCompile(`@([a-z][a-z0-9]*(?:[-\.][a-z0-9]+)*)`)
	issueRe   = regexp.MustCompile(`(?i)#([a-z][a-z0-9]*)-([a-z0-9]+)\b`)
)

// GetProjectName returns the basename for the project root.
func GetProjectName(projectRoot string) string {
	return filepath.Base(projectRoot)
}

// FormatMessage formats a message for display.
func FormatMessage(msg types.Message, projectName string, agentBases map[string]struct{}) string {
	return formatMessageWithOptions(msg, projectName, agentBases, true, nil)
}

// FormatMessageFull formats a message without truncation (for anchors).
func FormatMessageFull(msg types.Message, projectName string, agentBases map[string]struct{}) string {
	return formatMessageWithOptions(msg, projectName, agentBases, false, nil)
}

// FormatMessageWithQuote formats a message with an optional quoted message for inline display.
func FormatMessageWithQuote(msg types.Message, projectName string, agentBases map[string]struct{}, quotedMsg *types.Message) string {
	return formatMessageWithOptions(msg, projectName, agentBases, true, quotedMsg)
}

func formatMessageWithOptions(msg types.Message, projectName string, agentBases map[string]struct{}, truncate bool, quotedMsg *types.Message) string {
	editedSuffix := ""
	if msg.Edited || msg.EditCount > 0 || msg.EditedAt != nil {
		editedSuffix = " (edited)"
	}
	idBlock := fmt.Sprintf("%s[%s#%s%s %s]%s", dim, bold, projectName, reset, dim+msg.ID+editedSuffix, reset)

	// Check for answer message format
	if strings.HasPrefix(msg.Body, "answered @") {
		return formatAnswerMessage(msg, projectName, idBlock, agentBases)
	}

	color := getAgentColor(msg.FromAgent, msg.Type, nil)
	displayBody := msg.Body
	if truncate {
		displayBody = truncateForDisplay(msg.Body, msg.ID)
	}

	// Format quote block if present
	quoteBlock := ""
	if quotedMsg != nil {
		quoteBlock = formatQuoteBlock(quotedMsg, agentBases)
	}

	// Format reactions if present
	reactionSuffix := formatReactionSummary(msg.Reactions)
	if reactionSuffix != "" {
		reactionSuffix = " " + gray + "(" + reactionSuffix + ")" + reset
	}

	if color != "" {
		coloredBody := colorizeBody(displayBody, color, agentBases)
		coloredBody = highlightIssueIDs(coloredBody, color)
		if quoteBlock != "" {
			return fmt.Sprintf("%s %s@%s:%s\n%s\n\"%s\"%s%s", idBlock, color, msg.FromAgent, reset, quoteBlock, color+coloredBody, reset, reactionSuffix)
		}
		return fmt.Sprintf("%s %s@%s: \"%s\"%s%s", idBlock, color, msg.FromAgent, coloredBody, reset, reactionSuffix)
	}

	highlightedBody := highlightIssueIDs(highlightMentions(displayBody), "")
	if quoteBlock != "" {
		return fmt.Sprintf("%s @%s:\n%s\n\"%s\"%s", idBlock, msg.FromAgent, quoteBlock, highlightedBody, reactionSuffix)
	}
	return fmt.Sprintf("%s @%s: \"%s\"%s", idBlock, msg.FromAgent, highlightedBody, reactionSuffix)
}

// formatQuoteBlock formats the quoted message for inline display.
func formatQuoteBlock(quotedMsg *types.Message, _ map[string]struct{}) string {
	// Get first few lines of quoted message, truncated
	lines := strings.Split(quotedMsg.Body, "\n")

	// Take at most 3 lines, truncate each
	maxLines := 3
	if len(lines) > maxLines {
		lines = lines[:maxLines]
	}

	var quotedLines []string
	for _, line := range lines {
		if len(line) > 80 {
			line = line[:77] + "..."
		}
		quotedLines = append(quotedLines, gray+"> "+line+reset)
	}

	// Add source info
	shortID := quotedMsg.ID
	if len(shortID) > 12 {
		shortID = shortID[:12]
	}
	quotedLines = append(quotedLines, fmt.Sprintf("%s> [%s @%s]%s", gray, shortID, quotedMsg.FromAgent, reset))

	return strings.Join(quotedLines, "\n")
}

// formatAnswerMessage renders answer messages with Q&A colorization.
func formatAnswerMessage(msg types.Message, projectName, idBlock string, agentBases map[string]struct{}) string {
	lines := strings.Split(msg.Body, "\n")
	if len(lines) == 0 {
		return idBlock + " " + msg.Body
	}

	// Parse header: "answered @opus @designer"
	header := lines[0]
	askerPart := strings.TrimPrefix(header, "answered ")

	// Get colors
	answererColor := getAgentColor(msg.FromAgent, msg.Type, nil)
	if answererColor == "" {
		answererColor = reset
	}

	// Extract first asker for question color
	askerColor := gray
	if matches := mentionRe.FindStringSubmatch(askerPart); len(matches) > 1 {
		askerColor = getAgentColor(matches[1], types.MessageTypeAgent, nil)
		if askerColor == "" {
			askerColor = gray
		}
	}

	// Build formatted output
	var out strings.Builder

	// Header: @answerer answered @asker:
	out.WriteString(fmt.Sprintf("%s %s@%s%s answered %s:\n", idBlock, answererColor, msg.FromAgent, reset, askerPart))

	// Parse Q&A blocks
	inAnswer := false
	boldWhite := ansiCode("\x1b[1;37m")

	for i := 1; i < len(lines); i++ {
		line := lines[i]

		if strings.HasPrefix(line, "Q: ") {
			inAnswer = false
			questionText := strings.TrimPrefix(line, "Q: ")
			out.WriteString(fmt.Sprintf("%sQ:%s %s%s%s\n", boldWhite, reset, askerColor, questionText, reset))
		} else if strings.HasPrefix(line, "A: ") {
			inAnswer = true
			answerText := strings.TrimPrefix(line, "A: ")
			out.WriteString(fmt.Sprintf("%sA:%s %s%s%s\n", boldWhite, reset, answererColor, answerText, reset))
		} else if strings.TrimSpace(line) == "" {
			out.WriteString("\n")
		} else if inAnswer && strings.HasPrefix(line, "   ") {
			// Continuation of answer
			out.WriteString(fmt.Sprintf("%s%s%s\n", answererColor, line, reset))
		} else {
			out.WriteString(line + "\n")
		}
	}

	return strings.TrimRight(out.String(), "\n")
}

func ansiCode(code string) string {
	if noColor {
		return ""
	}
	return code
}

func getAgentColor(agentID string, msgType types.MessageType, colorMap map[string]int) string {
	if noColor {
		return ""
	}
	if msgType == types.MessageTypeUser {
		return ""
	}
	if agentID == "system" {
		return gray
	}

	parsed, err := core.ParseAgentID(agentID)
	if err != nil {
		index := hashString(agentID) % len(colorPairs)
		return colorPairs[index].Bright
	}

	base := parsed.Base
	colorIndex := hashString(base) % len(colorPairs)
	if colorMap != nil {
		if mapped, ok := colorMap[base]; ok {
			colorIndex = mapped % len(colorPairs)
		}
	}

	version := 1
	if parsed.Version != nil {
		version = *parsed.Version
	}

	pair := colorPairs[colorIndex]
	if version%2 == 1 {
		return pair.Bright
	}
	return pair.Dim
}

func hashString(value string) int {
	hash := 0
	for i := 0; i < len(value); i++ {
		hash = ((hash << 5) - hash) + int(value[i])
	}
	if hash < 0 {
		return -hash
	}
	return hash
}

func highlightIssueIDs(body, senderColor string) string {
	if noColor {
		return body
	}
	return issueRe.ReplaceAllStringFunc(body, func(match string) string {
		return issueBG + whiteText + match + reset + senderColor
	})
}

func highlightMentions(body string) string {
	if noColor {
		return body
	}
	matches := mentionRe.FindAllStringSubmatchIndex(body, -1)
	if len(matches) == 0 {
		return body
	}

	var out strings.Builder
	last := 0
	for _, match := range matches {
		if len(match) < 2 {
			continue
		}
		start, end := match[0], match[1]
		if start > 0 {
			prev, _ := utf8.DecodeLastRuneInString(body[:start])
			if isAlphaNum(prev) {
				continue
			}
		}
		out.WriteString(body[last:start])
		out.WriteString(cyan)
		out.WriteString(body[start:end])
		out.WriteString(reset)
		last = end
	}
	out.WriteString(body[last:])
	return out.String()
}

func colorizeBody(body, senderColor string, agentBases map[string]struct{}) string {
	if noColor || senderColor == "" {
		return body
	}

	matches := mentionRe.FindAllStringSubmatchIndex(body, -1)
	if len(matches) == 0 {
		return body
	}

	var out strings.Builder
	last := 0
	for _, match := range matches {
		if len(match) < 4 {
			continue
		}
		start, end := match[0], match[1]
		nameStart, nameEnd := match[2], match[3]
		if start > 0 {
			prev, _ := utf8.DecodeLastRuneInString(body[:start])
			if isAlphaNum(prev) {
				continue
			}
		}

		out.WriteString(body[last:start])
		name := body[nameStart:nameEnd]
		replacement := body[start:end]
		lower := strings.ToLower(name)

		switch {
		case lower == "all":
			if mentionColor := getAgentColor(name, types.MessageTypeAgent, nil); mentionColor != "" {
				replacement = mentionColor + replacement + senderColor
			}
		case strings.Contains(name, "."):
			if mentionColor := getAgentColor(name, types.MessageTypeAgent, nil); mentionColor != "" {
				replacement = mentionColor + replacement + senderColor
			}
		case agentBases != nil:
			if _, ok := agentBases[lower]; ok {
				if mentionColor := getAgentColor(name, types.MessageTypeAgent, nil); mentionColor != "" {
					replacement = mentionColor + replacement + senderColor
				}
			} else if len(name) >= 3 && len(name) <= 15 && !strings.Contains(name, ".") {
				replacement = bold + replacement + reset + senderColor
			}
		default:
			if len(name) >= 3 && len(name) <= 15 {
				if mentionColor := getAgentColor(name, types.MessageTypeAgent, nil); mentionColor != "" {
					replacement = mentionColor + replacement + senderColor
				}
			}
		}

		out.WriteString(replacement)
		last = end
	}
	out.WriteString(body[last:])
	return out.String()
}

func truncateForDisplay(body, msgID string) string {
	lines := strings.Split(body, "\n")
	if len(lines) <= maxDisplayLines {
		return body
	}

	truncated := strings.Join(lines[:maxDisplayLines], "\n")
	remaining := len(lines) - maxDisplayLines
	return fmt.Sprintf("%s\n... (%d more lines. Use 'fray get %s' to see full)", truncated, remaining, msgID)
}

func isAlphaNum(r rune) bool {
	return unicode.IsLetter(r) || unicode.IsDigit(r)
}

// FormatMessagePreview formats a message as a one-line preview for accordion display.
// Format: "  [msg-abc123] @agent: First line of message..."
func FormatMessagePreview(msg types.Message, projectName string) string {
	// Get first line of body
	firstLine := strings.Split(msg.Body, "\n")[0]

	// Truncate if too long
	maxLen := 50
	if len(firstLine) > maxLen {
		firstLine = firstLine[:maxLen] + "..."
	}

	shortID := msg.ID
	if len(shortID) > 12 {
		shortID = shortID[:12]
	}

	return fmt.Sprintf("  %s[%s]%s @%s: %s", dim, shortID, reset, msg.FromAgent, firstLine)
}

// AccordionOptions configures accordion behavior.
type AccordionOptions struct {
	Threshold    int  // Show accordion if more than this many messages (0 = use default)
	HeadCount    int  // Messages to show at start (0 = use default)
	TailCount    int  // Messages to show at end (0 = use default)
	ShowAll      bool // If true, disable accordion and show all messages
	ProjectName  string
	AgentBases   map[string]struct{}
	QuotedMsgs   map[string]*types.Message // Map of message ID -> quoted message for inline display
}

// FormatMessageListAccordion formats a list of messages with accordion collapsing.
// Returns a slice of formatted lines ready to print.
func FormatMessageListAccordion(messages []types.Message, opts AccordionOptions) []string {
	if len(messages) == 0 {
		return nil
	}

	threshold := opts.Threshold
	if threshold == 0 {
		threshold = DefaultAccordionThreshold
	}
	headCount := opts.HeadCount
	if headCount == 0 {
		headCount = AccordionHeadCount
	}
	tailCount := opts.TailCount
	if tailCount == 0 {
		tailCount = AccordionTailCount
	}

	// Helper to format a message with its quote if available
	formatMsg := func(msg types.Message) string {
		var quotedMsg *types.Message
		if msg.QuoteMessageGUID != nil && opts.QuotedMsgs != nil {
			quotedMsg = opts.QuotedMsgs[*msg.QuoteMessageGUID]
		}
		return formatMessageWithOptions(msg, opts.ProjectName, opts.AgentBases, true, quotedMsg)
	}

	// If ShowAll or under threshold, format all messages normally
	if opts.ShowAll || len(messages) <= threshold {
		lines := make([]string, len(messages))
		for i, msg := range messages {
			lines[i] = formatMsg(msg)
		}
		return lines
	}

	// Accordion: head + collapsed middle + tail
	var lines []string

	// Head messages (full format)
	for i := 0; i < headCount && i < len(messages); i++ {
		lines = append(lines, formatMsg(messages[i]))
	}

	// Middle messages (preview format)
	middleStart := headCount
	middleEnd := len(messages) - tailCount
	if middleEnd > middleStart {
		collapsedCount := middleEnd - middleStart
		lines = append(lines, fmt.Sprintf("%s  ... %d messages collapsed ...%s", dim, collapsedCount, reset))
		for i := middleStart; i < middleEnd; i++ {
			lines = append(lines, FormatMessagePreview(messages[i], opts.ProjectName))
		}
		lines = append(lines, fmt.Sprintf("%s  ... end collapsed ...%s", dim, reset))
	}

	// Tail messages (full format)
	for i := middleEnd; i < len(messages); i++ {
		if i >= headCount { // Avoid duplicates if list is small
			lines = append(lines, formatMsg(messages[i]))
		}
	}

	return lines
}

// formatReactionSummary formats reactions for display.
// Single reaction: "👍 alice", multiple: "👍x3"
func formatReactionSummary(reactions map[string][]types.ReactionEntry) string {
	if len(reactions) == 0 {
		return ""
	}
	keys := make([]string, 0, len(reactions))
	for reaction := range reactions {
		keys = append(keys, reaction)
	}
	sort.Strings(keys)

	parts := make([]string, 0, len(keys))
	for _, reaction := range keys {
		entries := reactions[reaction]
		count := len(entries)
		if count == 0 {
			continue
		}
		if count == 1 {
			parts = append(parts, fmt.Sprintf("%s %s", reaction, entries[0].AgentID))
		} else {
			parts = append(parts, fmt.Sprintf("%sx%d", reaction, count))
		}
	}
	return strings.Join(parts, " · ")
}
